src/deploy/functions/params.ts

```diff
@@ -0,0 +1,332 @@
+}
+
+interface SelectOptions<T> {
+  // Optional human-facing value for this option (e.g. "US Central (Iowa)" instead of value
+  // "us-central1")
+  label?: string;
+
+  // Actual value of the parameter if this option is selected
+  value: T;
+}
+
+export interface SelectInput<T> {
+  type: "select";
+
+  select: Array<SelectOptions<T>>;
+}
+
+export type Param = StringParam | IntParam;
+type ParamValue = string | number | boolean;
+
+/**
+ * Calls the corresponding resolveX function for the type of a param.
+ * To be used when resolving the default value of a param, if CEL.
+ * It's an error to call this on a CEL expression that depends on params not already known in the currentEnv.
+ */
+function resolveDefaultCEL(
+  type: string,
+  expr: CEL,
+  currentEnv: Record<string, ParamValue>
+): ParamValue {
+  const deps = dependenciesCEL(expr);
+  const allDepsFound = deps.every((dep) => {
+    return Object.keys(currentEnv).includes(dep);
+  });
+  if (!allDepsFound) {
+    throw new FirebaseError(
+      "Build specified parameter with un-resolvable default value " +
+        expr +
+        "; dependencies missing."
+    );
+  }
+
+  switch (type) {
+    case "string":
+      return resolveString(expr, currentEnv);
+    case "int":
+      return resolveInt(expr, currentEnv);
+    default:
+      throw new FirebaseError(
+        "Build specified parameter with default " + expr + " of unsupported type"
+      );
+  }
+}
+/**
+ * Tests whether a mooted ParamValue literal is of the correct type to be the value for a Param.
+ */
+function canSatisfyParam(param: Param, value: ParamValue): boolean {
+  if (param.type === "string") {
+    return typeof value === "string";
+  } else if (param.type === "int") {
+    return typeof value === "number" && Number.isInteger(value);
+  }
+  assertExhaustive(param);
+}
+
+/**
+ * A param defined by the SDK may resolve to:
+ * - the value of a Cloud Secret in the same project with name == param name (not implemented yet), but only if it's a SecretParam
+ * - a literal value of the same type already defined in one of the .env files with key == param name
+ * - the value returned by interactively prompting the user
+ *   - the default value of the prompt comes from the SDK via param.default, which may be a literal value or a CEL expression
+ *   - if the default CEL expression is not resolvable--it depends on a param whose value is not yet known--we throw an error
+ *   - yes, this means that the same set of params may or may not throw depending on the order the SDK provides them to us in
+ *   - after prompting, the resolved value of the param is written to the most specific .env file available
+ */
+export async function resolveParams(
+  params: Param[],
+  projectId: string,
+  userEnvs: Record<string, ParamValue>
+): Promise<Record<string, ParamValue>> {
+  const paramValues: Record<string, ParamValue> = {};
+
+  for (const param of params.filter((param) => userEnvs.hasOwnProperty(param.param))) {
```

Comment:
FYI, you're calling filter twice with inverted fields. There's a tool for this in `src/functional.ts`. You could theoretically say:

```
const [resolved, outstanding] = functional.partition(params, (param) => userEnvs.hasOwnProperty(param.param));

const typeMismatch = resolved.find((param) => !canSatisfyParam(param, userEnvs[param.param]));
if (typeMismatch) {
  throw new Error(...);
}

for (const param of outstanding) {
  ...
}
```

This is just an FYI, not an instance to change.